Fix TextStream dumping of Objective-C types

smfr opened the pull request:
#### 71add0ad8094e5feccf1bec54608d1889f1be051
<pre>
Fix TextStream dumping of Objective-C types
<a href="https://bugs.webkit.org/show_bug.cgi?id=262160">https://bugs.webkit.org/show_bug.cgi?id=262160</a>
<a href="https://rdar.apple.com/116094853">rdar://116094853</a>

Reviewed by Richard Robinson.

Having both `operator&lt;&lt;(id)` and `operator&lt;&lt;(NSArray *)` caused problems, specifically
trying to dump a Class would hit the latter for some reason.

Remove ambiguity by removing `operator&lt;&lt;(NSArray *)` and have all dumping of Objective-C
types go through `operator&lt;&lt;(id)` which internally queries the object type. Add support
for dumping `Class` and `NSDictionary`.

Move functions that dump CG types out of TextStream and into TextStreamCocoa.h as standalone
functions, so that platform-independent headers aren&apos;t polluted with CG types.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/text/TextStream.cpp:
* Source/WTF/wtf/text/TextStream.h:
* Source/WTF/wtf/text/TextStreamCocoa.h: Added.
* Source/WTF/wtf/text/cocoa/TextStreamCocoa.mm:
(WTF::TextStream::operator&lt;&lt;):
(WTF::operator&lt;&lt;):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/TextStreamCocoa.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/270118@main">https://commits.webkit.org/270118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d8eb604a1b0093e0a1fed9d706bb560676b2240

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26710 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22589 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24856 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/570 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22976 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24832 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2180 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21230 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27293 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1907 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22159 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28344 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21373 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22411 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22502 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26142 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/23858 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/166 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/31267 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3152 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6860 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5897 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2299 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/31236 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2214 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6534 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WTF/wtf/text/cocoa/TextStreamCocoa.mm:
```diff
@@ -19,54 +19,79 @@
+
+    if ([object isKindOfClass:[NSDictionary class]]) {
+        outputDictionary(object);
+        return ts;
```
is this line needed, since the line underneath will return ts anyways?

smfr replied:
I think the explicit early return is good (someone might add another clause below this one).

smfr committed:
Fix TextStream dumping of Objective-C types
https://bugs.webkit.org/show_bug.cgi?id=262160
rdar://116094853

Reviewed by Richard Robinson.

Having both `operator<<(id)` and `operator<<(NSArray *)` caused problems, specifically
trying to dump a Class would hit the latter for some reason.

Remove ambiguity by removing `operator<<(NSArray *)` and have all dumping of Objective-C
types go through `operator<<(id)` which internally queries the object type. Add support
for dumping `Class` and `NSDictionary`.

Move functions that dump CG types out of TextStream and into TextStreamCocoa.h as standalone
functions, so that platform-independent headers aren't polluted with CG types.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/text/TextStream.cpp:
* Source/WTF/wtf/text/TextStream.h:
* Source/WTF/wtf/text/TextStreamCocoa.h: Added.
* Source/WTF/wtf/text/cocoa/TextStreamCocoa.mm:
(WTF::TextStream::operator<<):
(WTF::operator<<):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/TextStreamCocoa.mm:
(TEST):

Canonical link: https://commits.webkit.org/270118@main
